### PR TITLE
shared: Add CRC64NVME checksum type support

### PR DESCRIPTION
**Pull request by nl0**
## Summary
Add CRC64NVME checksum type and algorithm parameter to quilt_shared for priority-based checksum selection.

## Changes
- Extend `ChecksumType` enum with `CRC64NVME` (uppercase for S3 API compatibility)
- Add `Checksum.crc64nvme()` helper method for base64-encoded CRC64NVME values
- Add `Checksum.empty_crc64nvme()` cached constant for empty file checksums
- Add `checksum_algorithm` parameter to `S3HashLambdaParams` (defaults to SHA256_CHUNKED for backward compatibility)

## Context
CRC64NVME is AWS S3's default checksum algorithm (since December 2024). These changes enable priority-based checksum selection and leveraging precomputed checksums from S3 for better performance.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h2>Greptile Overview</h2>

<h3>Greptile Summary</h3>

This PR adds CRC64NVME checksum support to enable priority-based checksum selection and leverage AWS S3's native checksum algorithm (default since December 2024).

## Key Changes
- Added `ChecksumAlgorithm` enum for algorithm selection (SHA256_CHUNKED, CRC64NVME)
- Added `ChecksumType.CRC64NVME` enum value for manifest serialization
- Implemented efficient CRC64-NVME combining using GF(2) matrix multiplication with O(log n) complexity
- Added `Checksum.crc64nvme()`, `crc64nvme_from_parts()`, and `empty_crc64nvme()` helper methods
- Added `checksum_algorithm` parameter to `S3HashLambdaParams` and `S3CopyLambdaParams`
- Updated boto3 type stubs from `boto3-stubs` to `types-boto3` across multiple files
- Bumped quilt3 minimum version from ~5.4 to >=7,<8

## Issues Found
- **Critical**: `Checksum.get_empty()` and `Checksum.from_s3_checksum()` methods missing exhaustive checks with `TX.assert_never()` - will return `None` for unhandled algorithm values instead of raising errors
- Previous thread noted concern about empty CRC64NVME being all zeros - this appears correct based on CRC64 properties, but should be verified against AWS S3 behavior

## Test Coverage
Comprehensive test suite added in `test_crc64.py` with known value validation, edge cases, associativity tests, and performance validation.

<h3>Confidence Score: 3/5</h3>

- This PR has two critical bugs in `pkgpush.py` where methods can return `None` unexpectedly, but the implementation is otherwise solid with good test coverage
- Score reflects well-designed CRC64 implementation with comprehensive tests and proper type safety in most places, but docked 2 points for two critical bugs that will cause runtime errors when methods return `None` instead of Checksum objects
- Pay close attention to `py-shared/src/quilt_shared/pkgpush.py` - two methods are missing exhaustive checks and will return `None` instead of raising errors

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| py-shared/src/quilt_shared/pkgpush.py | 2/5 | Adds CRC64NVME checksum support with new enum types and helper methods. Critical issues: two methods missing exhaustive checks will return None instead of raising errors. |
| py-shared/src/quilt_shared/crc64.py | 5/5 | New efficient CRC64-NVME implementation using GF(2) matrix multiplication. Well-tested with comprehensive test coverage. Implementation looks solid. |
| py-shared/tests/test_crc64.py | 5/5 | Comprehensive test suite for CRC64 functionality with known value validation, edge cases, and performance tests. Tests cover validation, associativity, and correctness. |
| py-shared/pyproject.toml | 5/5 | Updates boto3 stubs from boto3-stubs to types-boto3 and bumps quilt3 minimum version from ~5.4 to >=7,<8. Dependency updates align with package ecosystem changes. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client as Client Code
    participant Params as S3HashLambdaParams
    participant Algo as ChecksumAlgorithm
    participant Checksum as Checksum Class
    participant CRC64 as crc64 module
    participant S3 as S3 API

    Note over Client,S3: Checksum Algorithm Selection & Processing

    Client->>Params: Create with checksum_algorithm
    Params->>Algo: Use ChecksumAlgorithm.CRC64NVME or SHA256_CHUNKED
    
    alt Empty File
        Client->>Checksum: get_empty(algorithm)
        Checksum->>Checksum: Return empty_crc64nvme() or empty_sha256_chunked()
        Checksum-->>Client: Checksum object
    else Multi-Part Upload
        loop For each part
            S3-->>Client: Part checksum (base64)
        end
        Client->>Checksum: crc64nvme_from_parts(checksums, sizes)
        Checksum->>CRC64: combine_crc64nvme(part_crcs, part_sizes)
        loop For each part after first
            CRC64->>CRC64: crc64_extend(combined, part_size)
            CRC64->>CRC64: XOR with part CRC
        end
        CRC64-->>Checksum: Combined CRC64 bytes
        Checksum-->>Client: Checksum object
    else S3 Response Parsing
        S3-->>Client: Object checksum response
        Client->>Checksum: from_s3_checksum(algorithm, value)
        Checksum->>Checksum: Decode base64 and create checksum
        Checksum-->>Client: Checksum object
    end
    
    Client->>Algo: s3_checksum_algorithm property
    Algo-->>Client: "CRC64NVME" or "SHA256"
    
    Client->>Algo: s3_checksum_field property
    Algo-->>Client: "ChecksumCRC64NVME" or "ChecksumSHA256"
```

### CODE REVIEW

**Comment by drernie** [comment]
This is required. Don't know why it is still draft. 😿

File: py-shared/src/quilt_shared/pkgpush.py [hunk 1]
```diff
@@ -91,33 +112,108 @@ def __str__(self):
...
+        Args:
+            algorithm: ChecksumAlgorithm enum value
+
+        Returns:
+            Checksum for empty file
+        """
+        if algorithm is ChecksumAlgorithm.CRC64NVME:
+            return cls.empty_crc64nvme()
+        if algorithm is ChecksumAlgorithm.SHA256_CHUNKED:
+            return cls.empty_sha256_chunked()
```
[the remark this answers is not in the document]
Comment (drernie) [reply py-shared/src/quilt_shared/pkgpush.py:189]: Isn't that a good thing? What's the downside of returning None?

File: py-shared/src/quilt_shared/pkgpush.py [hunk 2]
```diff
@@ -57,28 +59,47 @@ def from_pk(cls, pk: PhysicalKey):
...
-    credentials: AWSCredentials
-    location: S3ObjectSource
-    target: S3ObjectDestination
+    @classmethod
+    def from_str(cls, value: str):
+        for v in cls:
+            if v.value == value:
+                return v
+
+        raise ValueError(f"Unsupported checksum algorithm: {value}")
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/pkgpush.py]: doesn't this work?


```suggestion
        return cls(value)
```

File: py-shared/src/quilt_shared/pkgpush.py [hunk 3]
```diff
@@ -57,28 +59,47 @@ def from_pk(cls, pk: PhysicalKey):
...
+            return "ChecksumCRC64NVME"
+        TX.assert_never(self)
 
 
 class ChecksumType(str, enum.Enum):
+    """Checksum type identifiers for manifest serialization (includes legacy types)."""
+
     SHA256 = "SHA256"  # legacy
     SHA256_CHUNKED = "sha2-256-chunked"
+    CRC64NVME = "CRC64NVME"  # AWS native, default for objects uploaded after Dec 2024
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/pkgpush.py:102]: IMO just "default" it's not quite correct

File: py-shared/src/quilt_shared/pkgpush.py [hunk 4]
```diff
@@ -91,33 +112,108 @@ def __str__(self):
...
         return f"{self.__class__.__name__}({self!s})"
 
-    @staticmethod
-    def hash_parts(parts: T.Sequence[bytes]) -> bytes:
-        return hashlib.sha256(b"".join(parts)).digest()
+    # Factory methods for creating checksums from raw bytes
 
     @classmethod
     def sha256(cls, value: bytes):
+        """Create SHA256 checksum from digest bytes (hex-encoded)."""
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/pkgpush.py:119]: IMO `hex-encoded` is a bit confusing 
at first I had impression that **input** is hex-encoded

File: py-shared/src/quilt_shared/pkgpush.py [hunk 5]
```diff
@@ -91,33 +112,108 @@ def __str__(self):
...
+
+    @classmethod
+    def crc64nvme_from_parts(cls, part_checksums: T.Sequence[bytes], part_sizes: T.Sequence[int]):
+        """Create CRC64NVME checksum from part checksums and sizes."""
+        combined = combine_crc64nvme(part_checksums, part_sizes)
+        return cls.crc64nvme(combined)
+
+    # Empty checksums
+    _EMPTY_SHA256 = hashlib.sha256().digest()
+    _EMPTY_CRC64NVME = b"\x00" * 8  # CRC64 of empty object is 0
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/pkgpush.py:152]: does it really make sense to have constants for them?

File: py-shared/src/quilt_shared/pkgpush.py [hunk 6]
```diff
@@ -196,6 +292,20 @@ class PackageConstructEntry(pydantic.v1.BaseModel):
     meta: T.Optional[T.Dict[str, T.Any]] = None
 
 
+class S3HashLambdaParams(pydantic.v1.BaseModel):
+    credentials: AWSCredentials
+    scratch_buckets: T.Dict[str, str]
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/pkgpush.py]:
```suggestion
    scratch_buckets: dict[str, str]
```

File: py-shared/src/quilt_shared/crc64.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
+"""Efficient CRC64-NVME checksum combination using GF(2) matrix multiplication.
+
+This module provides O(log n) CRC64 extension using precomputed transformation matrices.
+The naive bit-by-bit approach would take O(data_len × 8) time, which is prohibitively slow
```
Comment (sir-sigurd) [inline py-shared/src/quilt_shared/crc64.py:4]: > O(data_len × 8)

😐

File: py-shared/tests/test_crc64.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+def test_extend_large_offset():
+    """Handles multi-GiB offsets efficiently (O(log n) performance)."""
+    crc = 0x1234567890ABCDEF
+    large_offset = 5 * 1024**3  # 5 GiB (AWS S3 max part size)
+    result = crc64_extend(crc, large_offset)
+    assert 0 <= result < (1 << 64)
+
+
+def test_extend_validation():
```
Comment (sir-sigurd) [inline py-shared/tests/test_crc64.py:33]: parametrize?

File: py-shared/tests/test_crc64.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+def test_combine_order_matters():
+    """Part order affects the result."""
+    part_crcs = [b"\x00\x00\x00\x00\x00\x00\x00\x01", b"\x00\x00\x00\x00\x00\x00\x00\x02"]
+    part_sizes = [1024, 2048]
+
+    forward = combine_crc64nvme(part_crcs, part_sizes)
+    reversed = combine_crc64nvme([part_crcs[1], part_crcs[0]], [part_sizes[1], part_sizes[0]])
+
+    assert forward != reversed
```
Comment (sir-sigurd) [inline py-shared/tests/test_crc64.py]:
```suggestion
    reverse = combine_crc64nvme(list(reversed(part_crcs)), list(reversed(part_sizes)))

    assert forward != reverse
```

File: py-shared/tests/test_crc64.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def test_combine_mismatched_lengths():
+    """Mismatched inputs raise ValueError."""
+    with pytest.raises(ValueError, match="must have the same length"):
+        combine_crc64nvme([b"\x00" * 8, b"\x01" * 8], [1024])
+
+
+def test_combine_many_parts():
+    """Handles 640 parts efficiently (~5 GiB file)."""
```
Comment (sir-sigurd) [inline py-shared/tests/test_crc64.py:89]: how do we check it handles it efficiently?

File: py-shared/CHANGELOG.md [hunk 11]
```diff
@@ -17,6 +17,7 @@ where verb is one of
 
 ## Changes
 
+- [Added] Add CRC64NVME checksum type support ([#4623](https://github.com/quiltdata/quilt/pull/4623))
```
Comment (sir-sigurd) [inline py-shared/CHANGELOG.md:20]:
```suggestion
- [Changed] Switch from `boto3-stubs` to recommended `types-boto3` ([#4623](https://github.com/quiltdata/quilt/pull/4623))
- [Changed] Bump quilt3 dependency ([#4623](https://github.com/quiltdata/quilt/pull/4623))
- [Added] Add CRC64NVME checksum type support ([#4623](https://github.com/quiltdata/quilt/pull/4623))
```

**Review by sir-sigurd — APPROVED** [review]

File: py-shared/src/quilt_shared/crc64.py (hunk 7 above)
Comment (nl0) [reply py-shared/src/quilt_shared/crc64.py:4]: yeah, it was timing out combining checksums for 8mb chunks

File: py-shared/tests/test_crc64.py (hunk 10 above)
Comment (nl0) [reply py-shared/tests/test_crc64.py:89]: by ensuring it does not timeout lol. maybe we should add some timer idk to test executes under a second or smth. specific perf details are not that important, as long as it is "sensible"

File: py-shared/src/quilt_shared/crc64.py (hunk 7 above)
Comment (sir-sigurd) [reply py-shared/src/quilt_shared/crc64.py:4]: I mean `O(data_len × 8)` = `O(data_len)`

Comment (nl0) [reply py-shared/src/quilt_shared/crc64.py:4]: well, yep, i guess it was just trying to emphasize the bit-per-bit nature of the operation (you could easily fix that tho)
